[GLIB] always update the active uri of the frame

pgorszkowski-igalia opened the pull request:
#### 4a9e02aa3418af0c2117db77f0e7edd70955fa23
<pre>
[GLIB] always update the active uri of the frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=251256">https://bugs.webkit.org/show_bug.cgi?id=251256</a>

Reviewed by Carlos Garcia Campos.

It fixes the case when the uri has changed after first call of
webkit_frame_get_uri and then webkit_frame_get_uri is called again.
Without this fix the returned uri is the initial one not the actual one.

Test: I modified existing caillback for uri change and add there additional
asserts to test added functionality.

* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp:
(webkitFrameSetUri):
(webkit_frame_get_uri):
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFramePrivate.h:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
(uriChangedCallback):

Canonical link: <a href="https://commits.webkit.org/259999@main">https://commits.webkit.org/259999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/773027b921625da20884eee60d5e5268d1091317

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38861 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115211 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/114942 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109932 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6262 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98238 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111782 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95541 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40108 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94431 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27184 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81784 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8338 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28536 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8829 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5090 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14451 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48079 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10373 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3719 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp:
```diff
@@ -148,9 +156,6 @@ const gchar* webkit_frame_get_uri(WebKitFrame* frame)
 
-    if (frame->priv->uri.isNull())
-        frame->priv->uri = frame->priv->webFrame->url().string().utf8();
-
```
I think we should keep this, in case the frame is created after the loading events.

pgorszkowski-igalia replied:
So if we keep setting the uri here then we don't need to update it via webkitFrameSetUri, right? It will be updated for each time someone calls webkit_frame_get_uri (so we can say: "updating on demand")

carlosgcampos replied:
No, we keep the null check

pgorszkowski-igalia replied:
Ah, ok, I think I have got your point.

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -181,28 +181,40 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
-        webkitWebPageSetURI(m_webPage, getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader()));
+        const auto uri = getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader());
+        auto webKitFrame = webkitFrameGetOrCreate(&frame);
+        webkitFrameSetUri(webKitFrame, uri);
```
This only works for the main frame, we should update the uri for any frame. I don't think we want to create the WebKitFrame object just to set the uri, if it already exists we update the uri. What's the value of the frame url at this point? I think we could call webkitFrameUpdateURI() and there we always update the uri with the url value of the WebFrame object.

pgorszkowski-igalia replied:
You are right, I was concerned about using here webkitFrameGetOrCreate and create WebKitWebFrame object only for setting the uri as you mentioned. At this point the WebKitWebFrame does not exist and has to be created.

pgorszkowski-igalia replied:
So the idea here is to first check if WebKitWebFrame exists and only then update its uri, right?

carlosgcampos replied:
Right

pgorszkowski-igalia replied:
> What's the value of the frame url at this point?

unfortunately it is empty

> I think we could call webkitFrameUpdateURI() and there we always update the uri with the url value of the WebFrame object

what do you mean by webkitFrameUpdateURI? new function?

carlosgcampos replied:
It's empty because frame url always gets the document loader, not the provisional one. In that case we indeed need a SetURI.

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -134,15 +134,43 @@ static WebFrameMap& webFrameMap()
+
+    if (!url.isNull())
+        webkitFrameSetURI(wrapperPtr->webkitFrame(), url);
+
```
This could be moved to the WebKitFrame initialization.

carlosgcampos replied:
Are we sure this is correct? Since we don't know at what point of the load we are, maybe it wouold be better to check first if we have a document loader and if not try with the provisional one.

pgorszkowski-igalia replied:
Not sure here but most of the test failed if I put documentLoader as first. It seems that both loader exist in the same time but documentLoader contains empty url, and provisionalDocumentLoader contains expected one(at least in many tests)

pgorszkowski-igalia replied:
> This could be moved to the WebKitFrame initialization.

Done

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -134,15 +134,43 @@ static WebFrameMap& webFrameMap()
+
+static WebKitFrame* webkitFrameGetOrCreate(WebFrame* webFrame)
+{
+    if (auto webKitFrame = webkitFrameGet(webFrame))
```
auto*

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -134,15 +134,43 @@ static WebFrameMap& webFrameMap()
+static CString getDocumentLoaderURL(DocumentLoader* documentLoader)
+{
+    if (!documentLoader)
+        return { };
```
Can it be nullptr? I have an assert in current code.

pgorszkowski-igalia replied:
After move setting uri to WebKitWebFrame initialization I put assert back here

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -168,41 +196,60 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
-        if (!frame.isMainFrame())
-            return;
-        webkitWebPageSetURI(m_webPage, getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader()));
+        auto webKitFrame = webkitFrameGet(&frame);
```
auto*

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -168,41 +196,60 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
-        webkitWebPageSetURI(m_webPage, getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader()));
+        auto webKitFrame = webkitFrameGet(&frame);
+        auto coreFrame = frame.coreFrame();
+        if (webKitFrame && coreFrame) {
```
I don't think coreFrame can be nullptr, and we are using it without a check below

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -168,41 +196,60 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
+        auto coreFrame = frame.coreFrame();
+        if (webKitFrame && coreFrame) {
+            const auto uri = getDocumentLoaderURL(coreFrame->loader().provisionalDocumentLoader());
+            if (!uri.isNull())
```
Why do we check this only for frames but not for the page?

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp:
```diff
@@ -63,10 +64,27 @@ static void webkit_frame_class_init(WebKitFrameClass*)
     WebKitFrame* frame = WEBKIT_FRAME(g_object_new(WEBKIT_TYPE_FRAME, NULL));
     frame->priv->webFrame = webFrame;
+
+    webkitFrameSetURI(frame, getURL(webFrame));
```
No need to compare uris here, we can do `frame->priv->uri = getURL(webFrame);` directly.

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -171,6 +180,7 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
     static CString getDocumentLoaderURL(DocumentLoader* documentLoader)
     {
         ASSERT(documentLoader);
+
```
Why adding this extra line here?

carlosgcampos commented on Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
```diff
@@ -179,30 +189,46 @@ class PageLoaderClient final : public API::InjectedBundle::PageLoaderClient {
+            webkitFrameSetURI(webKitFrame, uri);
+
+        if (frame.isMainFrame())
+            webkitWebPageSetURI(m_webPage, uri);
```
I see now that we can avoid the uri creation if it's not main frame and it doesn't exist. So, we could do something like:

```
auto* webKitFrame = webkitFrameGet(&frame);
if (!webkitFrame && !frame.isMainFrame())
    return;

const auto uri = getDocumentLoaderURL(frame.coreFrame()->loader().provisionalDocumentLoader());
...
```

pgorszkowski-igalia committed:
[GLIB] always update the active uri of the frame
https://bugs.webkit.org/show_bug.cgi?id=251256

Reviewed by Carlos Garcia Campos.

It fixes the case when the uri has changed after first call of
webkit_frame_get_uri and then webkit_frame_get_uri is called again.
Without this fix the returned uri is the initial one not the actual one.

Test: I modified existing caillback for uri change and add there additional
asserts to test added functionality.

* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp:
(webkitFrameSetUri):
(webkit_frame_get_uri):
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFramePrivate.h:
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/WebExtensionTest.cpp:
(uriChangedCallback):

Canonical link: https://commits.webkit.org/259999@main
